interfaces/apparmor/backend.go

```diff
@@ -310,16 +310,12 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	if err := os.MkdirAll(dir, 0755); err != nil {
 		return fmt.Errorf("cannot create directory for apparmor profiles %q: %s", dir, err)
 	}
-	_, removed, errEnsure := osutil.EnsureDirStateGlobs(dir, []string{glob1, glob2}, content)
-	// NOTE: load all profiles instead of just the changed profiles.  We're
-	// relying on apparmor cache to make this efficient. This gives us
-	// certainty that each call to Setup ends up with working profiles.
-	all := make([]string, 0, len(content))
-	for name := range content {
-		all = append(all, name)
-	}
-	sort.Strings(all)
-	errReload := reloadProfiles(all, dir, cache)
+	changed, removed, errEnsure := osutil.EnsureDirStateGlobs(dir, []string{glob1, glob2}, content)
+	// NOTE: we are now reloading only the changed profiles. In the past we
+	// used to reload all the profiles but this increased memory pressure due
+	// to a kernel bug.  https://bugs.launchpad.net/apparmor/+bug/1750594
```

Comment:
Shouldn't we still regenerate **all** profiles if SystemKey changes? Perhaps we should pass a flag down to backend's Setup from InterfaceManager.initialize()?

Reply:
The input to the system key is also an input to the profiles. If they have to change internally (because computed something and decided they need to have different plaintext) then this will be reflected by `changed` in this patch. If they have to change externally (because include statements expand to different text) then this will be already done at this point, because the apparmor init script compiles and loads all profiles.

At this point, the system is checking if anything _actually_ needs to be changed on disk and loaded into the kernel.

Reply:
"If they have to change externally (because include statements expand to different text) then this will be already done at this point, because the apparmor init script compiles and loads all profiles."

This isn't strictly true. Consider a system that has snapd and no NFS mounts. The user adds an NFS mount, then reboots. AppArmor runs first, loads the profiles and there is no include for NFS. snapd starts, adds the include, and now does nothing.

Reply:
Also, I'm not excited about this PR because we want to keep the profile load logic simple. @jrjohansen is working on the kernel patch with critical priority right now.

Reply:
When snapd starts and adds the NFS snippet it will make all profiles different and then will reload all (changed) profiles.

Reply:
@jdstrand this leak affects all kernels since 4.13.

4.4 -> not affected
4.8 -> not affected
4.10 -> not affected
4.13 -> affected
4.15 -> affected

mainline is not strongly affected though some memory leaks (less than in ubuntu kernels)

Reply:
Thank you for all the investigation and this PR, but I maintain that we should not work around this in the code of snapd. We should fix the kernels and workaround in the testsuite until those are fixed. As mentioned, @jrjohansen is working on this with priority.